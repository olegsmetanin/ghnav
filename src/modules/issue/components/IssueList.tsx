import * as React from 'react'

import { IIssue, IIssueListQuery } from 'interfaces/issue'
import { Theme, WithStyles, createStyles, withStyles } from '@material-ui/core/styles'

import Button from '@material-ui/core/Button'
import Grid from '@material-ui/core/Grid'
import { IProcessState } from 'interfaces/redux'
import { IssueListItem } from 'modules/issue/components/IssueListItem'
import { Loader } from 'common/components/Loader'

const styles = (theme: Theme) =>
  createStyles({
    gridItem: {
      marginBottom: theme.spacing.unit * 2
    },
    submitButtonWrap: {
      margin: 'auto',
      marginBottom: theme.spacing.unit * 2
    }
  })

export interface IBaseIssueListProps {
  process?: IProcessState
  query?: IIssueListQuery
  items?: IIssue[]

  onLoad: (query) => void
}

export class BaseIssueList extends React.Component<IBaseIssueListProps & WithStyles<typeof styles>, {}> {
  handleOnLoadMore = () => {
    const { query } = this.props
    const newQuery = { ...query, page: +query.page + 1, add: true }
    this.props.onLoad(newQuery)
  }

  render() {
    const { classes, items, process, query } = this.props

    return (
      <Grid container>
        {items &&
          items.map(value => {
            return (
              <Grid item key={value.id} xs={12} className={classes.gridItem}>
                <IssueListItem value={value} owner={query.owner} repo={query.repo} />
              </Grid>
            )
          })}
        {items && (
          <div key="button" className={classes.submitButtonWrap}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              onClick={this.handleOnLoadMore}
              disabled={process && process.isLoading}
            >
              {process && process.isLoading ? 'Loading...' : 'Load more'}
            </Button>
          </div>
        )}
        {!items && <Loader />}
      </Grid>
    )
  }
}

export const IssueList = withStyles(styles)(BaseIssueList)
