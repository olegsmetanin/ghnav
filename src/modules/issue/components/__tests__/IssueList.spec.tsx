/* eslint-env jest */
import { IssueList } from '../IssueList'
import React from 'react'
import renderer from 'react-test-renderer'

describe('IssueList', () => {
  it('renders correct', async () => {
    const props = {
      owner: 'owner',
      repo: 'repo',
      number: '1',
      page: 1,
      per_page: 10,
      items: [
        {
          id: '12',
          number: '12',
          // markdown
          body: 'I have iPhone 5s, 6, 7, 8, XS',
          title: 'title',
          create_at: 'create_at'
        }
      ],
      query: {
        owner: 'owner',
        repo: 'repo',
        number: 'number',
        page: 1,
        per_page: 10,
        filter: {
          state: 'all'
        }
      },
      process: {},
      onLoad: () => {
        return
      }
    }
    const component = renderer.create(<IssueList {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders button in progress', async () => {
    const props = {
      owner: 'owner',
      repo: 'repo',
      number: 'number',
      page: 1,
      per_page: 10,
      items: [
        {
          id: '12',
          number: '12',
          // markdown
          body: 'I have iPhone 5s, 6, 7, 8, XS',
          title: 'title',
          create_at: 'create_at'
        }
      ],
      query: {
        owner: 'owner',
        repo: 'repo',
        number: 'number',
        page: 1,
        per_page: 10,
        filter: {
          state: 'all'
        }
      },
      process: {
        isLoading: true
      },
      onLoad: () => {
        return
      }
    }
    const component = renderer.create(<IssueList {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
