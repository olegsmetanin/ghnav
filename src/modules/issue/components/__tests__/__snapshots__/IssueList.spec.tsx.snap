// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueList renders button in progress 1`] = `
<div
  className="MuiGrid-container-3"
>
  <div
    className="MuiGrid-item-4 MuiGrid-grid-xs-12-43 BaseIssueList-gridItem-1"
  >
    <div>
      <div
        className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 BaseIssueListItem-paper-100"
      >
        <div>
          <a
            href="/issue?owner=owner&repo=repo&num=12"
            onClick={[Function]}
          >
            <h6
              className="MuiTypography-root-128 MuiTypography-h6-145 MuiTypography-colorInherit-157"
            >
              #
              12
               
              title
               
            </h6>
          </a>
        </div>
        <div />
      </div>
    </div>
  </div>
  <div
    className="BaseIssueList-submitButtonWrap-2"
  >
    <button
      className="MuiButtonBase-root-189 MuiButtonBase-disabled-190 MuiButton-root-163 MuiButton-contained-174 MuiButton-containedPrimary-175 MuiButton-raised-177 MuiButton-raisedPrimary-178 MuiButton-disabled-183"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="submit"
    >
      <span
        className="MuiButton-label-164"
      >
        Loading...
      </span>
    </button>
  </div>
</div>
`;

exports[`IssueList renders correct 1`] = `
<div
  className="MuiGrid-container-3"
>
  <div
    className="MuiGrid-item-4 MuiGrid-grid-xs-12-43 BaseIssueList-gridItem-1"
  >
    <div>
      <div
        className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 BaseIssueListItem-paper-100"
      >
        <div>
          <a
            href="/issue?owner=owner&repo=repo&num=12"
            onClick={[Function]}
          >
            <h6
              className="MuiTypography-root-128 MuiTypography-h6-145 MuiTypography-colorInherit-157"
            >
              #
              12
               
              title
               
            </h6>
          </a>
        </div>
        <div />
      </div>
    </div>
  </div>
  <div
    className="BaseIssueList-submitButtonWrap-2"
  >
    <button
      className="MuiButtonBase-root-189 MuiButton-root-163 MuiButton-contained-174 MuiButton-containedPrimary-175 MuiButton-raised-177 MuiButton-raisedPrimary-178"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-164"
      >
        Load more
      </span>
      <span
        className="MuiTouchRipple-root-192"
      />
    </button>
  </div>
</div>
`;
